x-pack/plugins/alerting/server/saved_objects/migrations.ts

```diff
@@ -750,6 +757,39 @@ function addSecuritySolutionAADRuleTypeTags(
     : doc;
 }
 
+function stripOutRuntimeFieldsInOldESQuery(
+  doc: SavedObjectUnsanitizedDoc<RawRule>
+): SavedObjectUnsanitizedDoc<RawRule> {
+  const isESDSLrule =
+    isEsQueryRuleType(doc) && !isSerializedSearchSource(doc.attributes.params.searchConfiguration);
+
+  if (isESDSLrule) {
+    try {
+      const parsedQuery = JSON.parse(doc.attributes.params.esQuery as string);
+      // parsing and restringifying will cause us to lose the formatting so we only do so if this rule has
+      // fields other than `query` which is the only valid field at this stage
+      const hasFieldsOtherThanQuery = Object.keys(parsedQuery).some((key) => key !== 'query');
+      return hasFieldsOtherThanQuery
+        ? {
+            ...doc,
+            attributes: {
+              ...doc.attributes,
+              params: {
+                ...doc.attributes.params,
+                esQuery: JSON.stringify(pick(parsedQuery, 'query'), null, 4),
+              },
+            },
+          }
+        : doc;
+    } catch (err) {
+      throw new Error(
```

Comment:
I think this is ok.  I was worried if there was a way for a user to set the query string to a non-JSON value, but we validate it: https://github.com/elastic/kibana/blob/b4ddb67224e1906ad7a53dfe098017738fb86d1f/x-pack/plugins/stack_alerts/server/alert_types/es_query/rule_type_params.ts#L92-L105

The params are weird here as the validation depends on searchType - I wonder if somehow you could use searchSource for the searchType, but still pass in invalid JSON esQuery and get it stored.  I think even in that case, the migration would ignore it - the user would need to change the searchType to queryDSL before the migration would hit it, and then they'd hit the validation changing the searchType.

But still, is it worth possibly failing migration over this?  Particularly in the future if the logic here gets more complicated (with searchType/searchSource).

Reply:
Yeah, I was torn on this - I don't think that failing to migrate an already broken rule (as it won't parse in the rule either) should fail the upgrade, but I also felt like silently allowing it to pass could cause problems - mostly thinking about the broken release where we allowed the failure to decrypt through the migration process.

Not 100% sure which way to go.
Any thoughts @kobelb and @mikecote ?

Reply:
I've decided to avoid failing the upgrade.

Since the failure to migrate a rule, at worst, causes the rule to break when it next runs - this seems like a reasonable risk when the alternative is to allow a corrupted rule from preventing Kibana upgrading.
An error is logged on upgrade with the exact ID and a warning that legacy runtime fields might not have been removed from this which would help us understand if this rule is then subsequently misbehaving.

Given that an unparsable rule is not likely to be running at all - I struggle to see a scenario where this breaks a healthy rule. Seems like the best balance.